neural_compressor/adaptor/torch_utils/awq.py

```diff
@@ -0,0 +1,435 @@
+        self.example_inputs = example_inputs
+        if example_inputs is None:
+            assert dataloader is not None, "datalaoder or example_inputs is required."
+            self.example_inputs = get_example_input(dataloader)
+        # Step 1: get hidden states and kwargs of first block. 
+        self.total_block_args, self.total_block_kwargs = get_hidden_states(
+            model, dataloader=dataloader, n_samples=n_samples, calib_func=calib_func
+        )
+        # Step 2: get block list and block prefix, number
+        self.block_prefix, self.block_num = get_block_prefix(model)
+        self.block_list = fetch_module(model, self.block_prefix)
+        self.bits = bits
+        self.group_size = group_size
+        self.scheme = scheme
+        self.sym_full_range = sym_full_range
+        self.weight_config = weight_config
+        self.model = model
+
+    def quantize(self, auto_scale=True, mse_range=True, folding=False, return_int=False):
+        """Execute AWQ quantization.
+
+        Args:
+            auto_scale (bool, optional): whether search scale. Defaults to True.
+            mse_range (bool, optional): whether search clip range. Defaults to True.
+            folding (bool, optional): whether only allow update scale when it can be fold 
+                                      to upper layer. Defaults to False.
+            return_int (bool, optional): whether return int dtype with WeightOnlyLinear. 
+                                         Defaults to False.
+
+        Returns:
+            model: quantized model
+        """
+        # Step 1: get absorbed module list per block, includes self-absorption
+        # block_absorb_dict is split per block, includes all absorb relationship.
+        # absorb_layer_dict is the inverse of block_absorb_dict for all blocks
+        self.block_absorb_dict, self.absorb_layer_dict = _get_absorb_per_block(
+            self.model, self.example_inputs, 
+            # for only mse_range, folding is useless.
+            folding = folding if auto_scale else False,
+            weight_config=self.weight_config,
+        )
+        # process per block
+        for i, module_list in self.block_absorb_dict.items():
+            logger.info(f"Processing block: {i+1}/{self.block_num}")
+            if len(module_list) == 0:
+                logger.info(f"No need to process this block.")
+                continue
+            # Step 1: fetch all input values of each linear for scale calculation
+            # use the first linear for QKV tuple
+            block_name = self.block_prefix + '.' + str(i)
+            block = fetch_module(self.model, block_name)
+            module_hook_config = {
+                v[0].split(block_name + '.')[1]: ['input'] for v in module_list
+            }
+            def block_calibration(model):
+                for args, kwargs in zip(self.total_block_args, self.total_block_kwargs):
+                    model(*args, **kwargs)
```

Comment:
awq does not need any grad, you can add no_grad for all the forward

Reply:
good point, thanks